source/darwin/nw_socket.c

```diff
@@ -1560,55 +1562,63 @@ static void s_process_listener_state_changed_task(struct aws_task *task, void *a
 
-    /* we're connected! */
-
-    if (state == nw_listener_state_waiting) {
-        AWS_LOGF_DEBUG(AWS_LS_IO_SOCKET, "id=%p handle=%p: listener waiting ", (void *)nw_socket, (void *)nw_listener);
-
-    } else if (state == nw_listener_state_failed) {
-        AWS_LOGF_DEBUG(AWS_LS_IO_SOCKET, "id=%p handle=%p: listener failed ", (void *)nw_socket, (void *)nw_listener);
-
-        AWS_LOGF_ERROR(
-            AWS_LS_IO_SOCKET,
-            "id=%p handle=%p: connection error %d",
-            (void *)nw_socket,
-            (void *)nw_listener,
-            crt_error_code);
-
-        s_lock_base_socket(nw_socket);
-        struct aws_socket *aws_socket = nw_socket->base_socket_synced_data.base_socket;
-        if (nw_socket->on_accept_started_fn) {
-            nw_socket->on_accept_started_fn(aws_socket, crt_error_code, nw_socket->listen_accept_started_user_data);
-        }
-        s_unlock_base_socket(nw_socket);
-
-    } else if (state == nw_listener_state_ready) {
+    AWS_LOGF_TRACE(
+        AWS_LS_IO_SOCKET,
+        "id=%p handle=%p: listener state changed to %d ",
+        (void *)nw_socket,
+        (void *)nw_listener,
+        state);
 
-        s_lock_base_socket(nw_socket);
-        struct aws_socket *aws_socket = nw_socket->base_socket_synced_data.base_socket;
-        if (aws_socket && status == AWS_TASK_STATUS_RUN_READY) {
-            AWS_FATAL_ASSERT(nw_socket->mode == NWSM_LISTENER);
-            aws_socket->local_endpoint.port = nw_listener_get_port(nw_socket->os_handle.nw_listener);
-            if (nw_socket->on_accept_started_fn) {
-                nw_socket->on_accept_started_fn(aws_socket, AWS_OP_SUCCESS, nw_socket->listen_accept_started_user_data);
-            }
+    switch (state) {
+        case nw_listener_state_failed: {
             AWS_LOGF_DEBUG(
                 AWS_LS_IO_SOCKET,
-                "id=%p handle=%p: listener on port %d ready ",
+                "id=%p handle=%p: listener failed with error %d",
                 (void *)nw_socket,
                 (void *)nw_listener,
-                aws_socket->local_endpoint.port);
-        }
-        s_unlock_base_socket(nw_socket);
+                crt_error_code);
 
-    } else if (state == nw_listener_state_cancelled) {
-        AWS_LOGF_TRACE(
-            AWS_LS_IO_SOCKET, "id=%p handle=%p: listener cancelled ", (void *)nw_socket, (void *)nw_listener);
+            s_lock_base_socket(nw_socket);
+            s_lock_socket_synced_data(nw_socket);
+            s_set_socket_state(nw_socket, listener_state_changed_args->socket, ERROR);
+            s_unlock_socket_synced_data(nw_socket);
+            struct aws_socket *aws_socket = nw_socket->base_socket_synced_data.base_socket;
+            if (nw_socket->on_accept_started_fn) {
+                nw_socket->on_accept_started_fn(aws_socket, crt_error_code, nw_socket->listen_accept_started_user_data);
+            }
+            s_unlock_base_socket(nw_socket);
```

Comment:
Is there any chance that something in the accept started callback could invoke something that tries to lock the socket base?

Reply:
The base socket is only modified(write) at two places:  
(1) socket creation/init : The lock is just inited, and would not hold at this time. 
(2) aws_socket_cleanup(): This is where the base socket get nulled out. We only lock the base socket if the function is not on the event loop to avoid dead lock. Use this callback as an example, if we call aws_socket_cleanup() in `on_accept_started_fn`, as it is on the event loop, we will not lock the base socket here.